platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/ReusableBucketPool.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2021-2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.merkledb.files.hashmap;
+
+import com.swirlds.virtualmap.VirtualKey;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/**
+ * HalfDiskHashMap buckets are somewhat expensive resources. Every bucket has an
+ * underlying byte buffer to store bucket data and metadata, and the number of
+ * buckets is huge. This class provides a bucket pool, so buckets can be reused
+ * rather than created on every read / write call.
+ *
+ * <p>Bucket pool is accessed from multiple threads:
+ * <ul>
+ *     <li>Transaction thread, when a key path is loaded from HDHM as a part of
+ *     get or getForModify call</li>
+ *     <li>Lifecycle thread, when updated bucket is written to disk in the end
+ *     of HDHM flushing</li>
+ *     <li>HDHM background bucket reading threads</li>
+ *     <li>Warmup (aka prefetch) threads</li>
+ * </ul>
+ *
+ * <p>If buckets were created, updated, and then released (marked as available
+ * for other threads) on a single thread, this class would be as simple as a
+ * single {@link ThreadLocal} object. This is not the case, unfortunately. For
+ * example, when HDHM background reading threads read buckets from disk, buckets
+ * are requested from the pool by {@link BucketSerializer} as a part of data
+ * file collection read call. Then buckets are updated and put to a queue, which
+ * is processed on a different thread, virtual pipeline (aka lifecycle) thread.
+ * Only after that buckets can be reused. This is why the pool is implemented as
+ * an array of buckets with fast concurrent read/write access from multiple
+ * threads.
+ */
+public class ReusableBucketPool<K extends VirtualKey> {
+
+    /** Default number of reusable buckets in this pool */
+    private static final int DEFAULT_POOL_SIZE = 8192;
+
+    /** Pool size */
+    private final int poolSize;
+
+    /** Buckets */
+    private final AtomicReferenceArray<Bucket<K>> buckets;
+    /**
+     * To avoid synchronization on a single object, every bucket has its own lock. The
+     * locks are used to wait, if the pool is empty (no available buckets), or to wake
+     * up the waiting thread, when a bucket is released back to the pool.
+     */
+    private final AtomicReferenceArray<Object> locks;
```

Comment:
Instead of Object, can we use actual locks here and reduce the overhead?  We could also then avoid the need for AtomicReference (just `List<ReentrantLock>`, for instance)

Reply:
I can't use just a list, even final, as elements from this list are accessed from multiple threads. As for locks, let me try and check how good/bad they are performance wise.

Reply:
You should be able to use a list if it's a concurrent list (e.g. `List<Lock> locks = new CopyOnWriteArrayList<>()`), but it's not a large difference to use the `AtomicReferenceArray<Lock>` instead.
Another alternate would be a `ConcurrentSkipListMap<Bucket, Lock>`, so the bucket and lock are more closely tied (while keeping iteration low cost), but not sure if that fits the reusability of buckets.